allow to configure port for docker

karfau opened the pull request:
**Type:** bug (according to the related issue)

The following has been addressed in the PR:

* [x] There is a related issue
* [ ] Unit or Functional tests are included in the PR
  - ~I wasn't sure how to verify the modified `Dockerfiie` works without any custom port config (just like it worked before defaulting to port 4873)~.
    - ~Should I be able to run `npm test` inside the container? (didn't work for me)~
    - ~Should I be able to visit http://localhost:4873 in the browser and see something? (didn't work for me)~
    - when knowing how to ensure docker build is working I could add a `docker-compose.test.yaml` to enable automatic testing on docker hub/ docker cloud. Should this go into a separate PR?
<!--
Our bots should ensure:
* [ ] The PR passes CI testing
-->

**Description:**
Resolves #213

juanpicado commented:
* **Should I be able to run npm test inside the container? (didn't work for me)**
No, there is no unit test for this
* **Should I be able to visit http://localhost:4873 in the browser and see something? (didn't work for me)**
It should work. It does for me
```bash
 jpicado$ docker run -it --rm --name verdaccio -p 4873:4873 verdaccio/local
 warn --- config file  - /verdaccio/conf/config.yaml
 warn --- http address - http://0.0.0.0:4873/ - verdaccio/2.2.1
 http <-- 200, user: undefined(172.17.0.1), req: 'GET /', bytes: 0/1409
 http <-- 200, user: undefined(172.17.0.1), req: 'GET /-/static/jquery.min.js', bytes: 0/33417
 http <-- 200, user: undefined(172.17.0.1), req: 'GET /-/static/main.css', bytes: 0/26190
 http <-- 200, user: undefined(172.17.0.1), req: 'GET /-/static/main.js', bytes: 0/13440
 http <-- 200, user: undefined(172.17.0.1), req: 'GET /-/static/fontello.woff?10872183', bytes: 0/2844
 http <-- 200, user: undefined(172.17.0.1), req: 'GET /-/logo', bytes: 0/3447
```
* **when knowing how to ensure docker build is working I could add a docker-compose.sut.yaml to enable automatic testing on docker hub/ docker cloud. Should this go into a separate PR?**
Could yo elaborate this a bit more?

karfau commented:
* **Should I be able to visit http://localhost:4873 in the browser and see something?**  
I tried to build the image from master branch locally by just using `docker build .`.
That should work, right? Will give it another try when back home.

* **when knowing how to ensure docker build is working I could add a docker-compose.sut.yaml to enable automatic testing on docker hub/ docker cloud. Should this go into a separate PR?**  
I was referring to the following approach: https://docs.docker.com/docker-cloud/builds/automated-testing/
but confused that this means to have `*.test.yaml` files containing a `sut` service that is used to execute the tests. I know a bit on how to set up this stuff, so if there are some "tests" e.g. like the functional ones if they would using the docker image, or just a test that we can curl the browser page) you can make sure no broken docker image gets released.

juanpicado commented:
About the `docker build .` it takes ±10 minutes in my computer. So do not lost the faith 👍  if it takes to long (https://github.com/verdaccio/verdaccio/pull/181#issuecomment-309280108)
*docker-compose.sut.yaml * Awesome !! I wasn't aware of that, looking forward to see it running.

karfau commented:
yes, build time of the image takes some time (I added a `.dockerignore` to improve that) , but that wasn't the problem. After I restarted my machine today I can visit the page and see something. :+1: 

No I'm going to test my changes :)

If you have any idea or hint on how to run the functional tests against the docker instance, please tell me.

karfau commented:
about the `.dockerignore`: would you prefer to have a separate PR for that or can I include it in my changes?

juanpicado commented:
@karfau  yes please 👍

karfau commented:
hey, there.
The changes as of now will allow to do the following:
a) all commands that worked before continue to work.
b) when **running** the container you can provide a different port to listen on like this:
```bash
docker run --env PORT=5000 -it --rm --name verdaccio -p 5000:5000 verdaccio/local
```
which looks a bit useless and error prone, because you need to know what was written in the config file and match with that to make it work.

So I guess this is not what has been requested in #213?
 I will take the clarifying conversation over there, so it sticks in case we drop this PR :)

juanpicado commented:
@karfau about the functional testing, give me more time I'll explain how they work.

karfau commented:
I rebased and squashed, I think this is it for this pull request.

From my understanding this doesn't really fix the issue as of its description,
as it is still not possible to use the config file to set the port used by docker.
But I think it shows, mostly by documenting, that this "feature" is not needed in case of docker.

I restructured the docker section in the README a bit to order them by most common use case.
Quite some docker stuff in there now, maybe reduce that to the minimum to get started and link to a DOCKER.md for all the extra details?

I wasn't sure what the next version will be, so I added `2.?.?` to the README, please put in the right value in case this gets merged.

karfau commented:
I have no clue about the failing test, if I can help with anything, please tell me what to do.

About running the functional tests against docker, maybe we can continue the conversation in a different ticket?

juanpicado commented on .dockerignore:
```diff
@@ -0,0 +1,26 @@
+# https://docs.docker.com/engine/reference/builder/#dockerignore-file
+
+# consider them hidden
+.*
```
I'm not sure if this affect the `npm install` because in other branches we are using babel and dot files are important.

karfau replied:
I see, before I came to this entry Ih had like 15 entries starting with a dot in the file,
then I read the article I mention in the file and changed it to this.

You can always exclude certain files like in gitignore:
`!.babelrc` 
Then it becomes a white list, instead of blacklisting every single "hidden" file. :)

juanpicado replied:
Ok prefect that's valid for me then 👍  I'll notify @Meeeeow he is working in  a ui branch.

karfau replied:
added an example to the file.

juanpicado commented:
@karfau thanks for the changes, but, please do atomic :atom: commits , otherwise in the future is going to be hard to track changes.
The tests were failing, it travis is having a hard week 🤕  and fails randomly. Fixed already. 
I'm not a super Docker expert, but the changes looks good to me. I'll need to test them first before merge.

juanpicado commented:
@karfau One small thing, now you are playing around with the Docker readme section, could you rebase and move it to a `docker.md` in the wiki section and reference it from `wiki/README.md` and the root `README.md`?

karfau commented:
> please do atomic :atom: commits , otherwise in the future is going to be hard to track changes.
sorry, doing squashing and rebasing before merging to master all the time at work, so didn't think a second that you might want to keep those commits. Thx for pointing it out.

> One small thing, now you are playing around with the Docker readme section, could you rebase and move it to a docker.md in the wiki section and reference it from wiki/README.md and the root README.md?
Not sure I what you mean with "rebase" in this case, but sure I can move that chapter to the wiki.
Should I leave some basics in the README and link to wiki for more details, or you want to get rid of the whole docker section there?

juanpicado commented:
> Not sure I what you mean with "rebase" in this case, but sure I can move that chapter to the wiki.
Should I leave some basics in the README and link to wiki for more details, or you want to get rid of the whole docker section there?

Take a look this section https://github.com/verdaccio/verdaccio/blob/master/CONTRIBUTING.md#5-rebase-and-push-changes

Basically you have to rebase your branch onto `master` in order to get the latest changes. The rebase move your commits a the end (https://www.atlassian.com/git/tutorials/rewriting-history#git-rebase). This is really helpful because in case of conflicts you resolved in your side before push your changes.

karfau commented:
Just saw your changes of adding wiki folder to the repo, after I added a docker page to the github wiki.
I hope this doesn't cause any trouble.
I just rebased this morning that's why I didn't expect any news from that side :)

Going to reabse and do as you requested.

Should I leave some basics in the README and link to wiki for more details, or you want to get rid of the whole docker section there?

karfau commented:
> Take a look at CONTRIBUTION.md

Should have done this before...

juanpicado commented:
> Just saw your changes of adding wiki folder to the repo, after I added a docker page to the github wiki.
I hope this doesn't cause any trouble.
Not all all 👍 
> Should I leave some basics in the README and link to wiki for more details, or you want to get rid of the whole docker section there?
Sure, just let a small docker section that points to the wiki. README is on my TOP 5 of conflicts. I wanna keep it small.

karfau commented:
Did the requested changes.
Should I delete the Docker page from the wiki?

juanpicado commented:
@karfau yes you can delete it

juanpicado commented:
Perfect great job ! 👍  merging !
